features/telemetry/adapters/nxos/src/main/java/org/opennms/netmgt/telemetry/adapters/nxos/NxosGpbParserUtil.java

```diff
@@ -0,0 +1,192 @@
+ * OpenNMS(R) is Copyright (C) 1999-2017 The OpenNMS Group, Inc.
+ *
+ * OpenNMS(R) is a registered trademark of The OpenNMS Group, Inc.
+ *
+ * OpenNMS(R) is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * OpenNMS(R) is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with OpenNMS(R).  If not, see:
+ *      http://www.gnu.org/licenses/
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.netmgt.telemetry.adapters.nxos;
+
+import java.util.List;
+import java.util.Objects;
+
+import org.opennms.netmgt.telemetry.adapters.nxos.proto.TelemetryBis;
+import org.opennms.netmgt.telemetry.adapters.nxos.proto.TelemetryBis.Telemetry;
+
+/**
+ * Utility to parse any key/value metric from gpb messages.
+ *
+ * Since there is no easy way to get key/value metrics from repeated data_gpbkv,
+ * this utility provides some methods to get values for those metrics without
+ * bothering about structure they may follow.
+ *
+ * @author cgorantla
+ */
+public class NxosGpbParserUtil {
+
+    /**
+     * Given a Telemetry Message and field name, get it's value as double.
+     * 
+     * @param   msg  {@link org.opennms.netmgt.telemetry.adapters.nxos.proto.TelemetryBis.Telemetry} message
+     * @param   name  field name
+     * @return  value field value
+     */
+    public static Double getValueAsDouble(TelemetryBis.Telemetry msg, String name) {
+        for (TelemetryBis.TelemetryField field : msg.getDataGpbkvList()) {
+            TelemetryBis.TelemetryField subField = findFieldWithName(field, name);
+            if (subField != null) {
+                return getValue(subField);
+            }
+        }
+        return Double.NaN;
+    }
+
+    /**
+     * Given a Telemetry Message and field name, get it's value as string.
+     * 
+     * @param   msg  {@link org.opennms.netmgt.telemetry.adapters.nxos.proto.TelemetryBis.Telemetry} message
+     * @param   name  field name
+     * @return  value field value
+     */
+    public static String getValueAsString(TelemetryBis.Telemetry msg, String name) {
+        for (TelemetryBis.TelemetryField field : msg.getDataGpbkvList()) {
+            TelemetryBis.TelemetryField subField = findFieldWithName(field, name);
+            if (subField != null) {
+                return subField.getStringValue();
```

Comment:
We should cast the actual value to a string here instead - similar to what we did the the double above.